streamingccl: add physical_replication.consumer.split_on_job_retry setting

msbutler opened the pull request:
This patch adds the default off
physical_replication.consumer.split_on_job_retry setting, which, when enabled
issues admin splits over the topology after a job level retry. This setting may
help c2c catch up more quickly after a replanning event, as it would prevent
the destination side from issueing admin splits during catchup scans.

Informs https://github.com/cockroachdb/cockroach/issues/114706

Release note: none

msbutler committed:
streamingccl: add physical_replication.consumer.split_on_job_retry setting

This patch adds the default off
physical_replication.consumer.split_on_job_retry setting, which, when enabled
issues admin splits over the topology after a job level retry triggered by
distSQL replanning. This setting may help c2c catch up more quickly after a
replanning event, as it would prevent the destination side from issueing admin
splits during catchup scans.

Informs #114706

Release note: none

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/114986)
<!-- Reviewable:end -->

msbutler commented on pkg/ccl/streamingccl/streamingest/stream_ingestion_dist.go:
@stevendanna I wonder if we want to add another gate on post initial splits to prevent them from occurring on every retry. We could only trigger them if the previous job level error involved replanning, or if we're resuming after pause, or if the hwm is sufficiently far behind.

stevendanna replied:
I like the idea of only resplitting if we also hit a replan since right now the replan is a little bit of a signal that our data distribution changed. It would also prevent us resplitting when we are in an unrelated error loop.

msbutler committed:
streamingccl: deflake TestStreamingReplanOnLag

The test would timeout under stress as it relies on node 1 falling behind, but
occasonially, partitionSpans would not include node 1 in the c2c distSQL plan.
This patch prevents this by modifying the CreateScatteredTable to wait until
the desired number of nodes have leases.

Release note: none

stevendanna commented:
Sorry I missed your follow-up here.

I believe in follow up discussions we were less clear on how helpful this is in general.  Happy to re-review if you are still interested in having this on master. But otherwise I think we could say we will will come back to this when we next address PCR bulk performance.
